### PR TITLE
test: introduced test harness

### LINKED ISSUE #304: Port Prism test harness
**User story.**
Port the existing Prism test harness once it has settled. At time of writing it is still being reworked and refactored, but once done, Spectral should get the same treatment.

**AC*
-Spectral uses the same test-harness library Prism does
- Run test-harness for every PR in CI
- Write some e2e smoke tests (up to 5 should be enough)

**Scope of Work**
- [ ] Extract the Prism harness utilities in a separate package
- [ ] Reinstall the utilities in Prism and Spectral

### PR DESCRIPTION
**Pull request by philsturgeon**
Introduced "test harness", a repository of simple files which contain description documents, commands to run, and their expected outputs or errors. This should help us avoid regression, and help us figure out how well built tools work on a variety of operating systems.

Fixes #304.

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

Add any other context about the pull request here. Remove this section if there is no additional context.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -134,7 +134,6 @@
   },
   "husky": {
     "hooks": {
-      "commit-msg": "commitlint -E HUSKY_GIT_PARAMS",
```
Comment (philsturgeon) [inline package.json]: Turning this off as we said in the handbook. You can still use `yarn commit` but its not required.

File: test-harness/index.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+          to: tmpFileHandle.name,
+        });
+
+        fs.writeFileSync(tmpFileHandle.name, scenario.document, { encoding: 'utf8' });
+      }
+    });
+
+    afterAll(() => {
+      if (scenario.document) {
+        tmpFileHandle.removeCallback(undefined, undefined, undefined, undefined);
```
Comment (lag-of-death) [inline test-harness/index.ts]: hm, so the `undefined`s are required here? catches the eye ;)

File: test-harness/oldcommandtests/__fixtures__/openapi-3.0-valid-custom.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+info:
+  version: 1.0.0
+  title: Stoplight
+  description: lots of text
+  contact:
+    name: fred
+servers:
+  - description: thing
+    url: http://localhost
+paths: {}
```
Comment (lag-of-death) [inline test-harness/oldcommandtests/__fixtures__/openapi-3.0-valid-custom.yaml]:
```suggestion
paths: {}

```

File: test-harness/index.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+const spectralBin = path.join(__dirname, '../cli-binaries/spectral-cli');
+
+interface IReplacement {
+  from: RegExp;
+  to: string;
+}
+
+function replaceVars(string: string, replacements: IReplacement[]) {
+  let result = string;
+  replacements.forEach(replace => {
```
[the remark this answers is not in the document]
Comment (lag-of-death) [reply test-harness/index.ts]: would be nice :+1:

[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/index.ts]: We are not accumulating a result we are making multiple passes replacing various different things in the same string.

File: test-harness/index.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+      const stderr = commandHandle.stderr.trim();
+      const stdout = commandHandle.stdout.trim();
+      const expectedStderr = replaceVars(scenario.stderr.trim(), replacements);
+      const expectedStdout = replaceVars(scenario.stdout.trim(), replacements);
+
+      if (expectedStderr) {
+        expect(stderr).toEqual(expectedStderr);
+      } else if (stderr) {
+        done(stderr);
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/index.ts]: done does help with async but it also lets you pass errors back. I can just throw it but this is also fine?

File: test-harness/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  /todos:
+    get:
+      responses:
+        200:
+          description: Get Todo Items
+          content:
+            'application/json':
+              example: hello
+====command====
+lint {document} | grep "something"
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/README.md]: Sure does. It doesn't care what the command is just something valid.

File: test-harness/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
+# Test Harness
+
+## Prerequisites
+
+* Install the project dependencies with `yarn`
+* Generate the binary for your platform with `yarn build.binary`. This will *also* compile the project from TS -> JS
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/README.md]: Seems to be?

File: test-harness/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (philsturgeon) [reply test-harness/index.ts]: Done

File: test-harness/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (P0lip) [reply test-harness/README.md]: Yeah, it's true. We need to compile some files.

File: test-harness/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (P0lip) [reply test-harness/index.ts]: `for of` FTW

File: tsconfig.json [hunk 8]
```diff
@@ -21,7 +21,7 @@
     "resolveJsonModule": true,
     "outDir": "dist",
     "sourceMap": true,
-    "declaration": true,
+    "declaration": false,
```
Comment (P0lip) [inline tsconfig.json]: This should be set to true.
```suggestion
    "declaration": true,
```

File: test-harness/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (P0lip) [reply test-harness/README.md]: Well, in fact it does not. I guess test harness works fine because we execute build somewhere on CI beforehand.

**Pull request merged**